Main perf tests

rawdaGastan opened the pull request:
### Issues

- https://github.com/threefoldtech/zos/issues/2050

### example output

[result.txt](https://github.com/threefoldtech/zos/files/12727833/result.txt)

Omarabdul3ziz committed:
init performance monitor package

Omarabdul3ziz committed:
handle new rmb call

Omarabdul3ziz committed:
resolve comments:
- introduce Task interface
- remove testing code
- use redis connection from pool
- unexpose some methods and fields
- update running zos node docs & add example for using perf

Omarabdul3ziz committed:
resolve comments:
- use redis pool correctly
- add GetAll method
- better naming for tasks with prefix `perf.`
- use unix date

Omarabdul3ziz committed:
update go.mod file

rawdaGastan committed:
merge

rawdaGastan committed:
add tcp, udp tests

rawdaGastan committed:
get random nodes to test

rawdaGastan committed:
add freefarm nodes

rawdaGastan commented:
waiting for 
- https://github.com/threefoldtech/zos/pull/2046

Omarabdul3ziz committed:
- add new handler for getAll method
- reuse Get code

Omarabdul3ziz committed:
Merge branch 'main' into main_perf_package

Omarabdul3ziz committed:
run tasks if it has no results stored, fix getAll method

rawdaGastan committed:
merge

rawdaGastan committed:
log error instead of making noded crash

muhamadazmy commented on pkg/graphql/graphql.go:
```diff
@@ -0,0 +1,68 @@
+// Package graphql for grid graphql support
+package graphql
```
Did we really needed to implement this from scratch ? can't we use an existing graphql client ?

muhamadazmy commented on cmds/modules/noded/main.go:
```diff
@@ -202,6 +205,41 @@ func action(cli *cli.Context) error {
 		return errors.Wrap(err, "failed to create a new perfMon")
 	}
 
+	if exists := iperf.Exists(zinit.Default()); exists {
```
As discussed before running a tests against public nodes happens from ALL nodes (even the ones without public config). The idea is:
- Public nodes running iperf running in test mode 
- All nodes run test against the nodes that has iperf server running 
- Nodes measure their up/down speed against the random public nodes.

muhamadazmy commented on cmds/modules/noded/main.go:
```diff
@@ -202,6 +205,41 @@ func action(cli *cli.Context) error {
+		if err != nil {
+			return errors.Wrap(err, "failed to list nodes from graphql")
+		}
+
```
This is too much to be done by noded. Noded all what it cares about is to register a certain test against the perf framework. 

The test ON RUNNING, will pick the nodes to test against. means the nodes can be different each time the test runs

muhamadazmy commented on cmds/modules/noded/main.go:
```diff
@@ -202,6 +205,41 @@ func action(cli *cli.Context) error {
+				TaskID:    "TCPTask",
+				Schedule:  "* */5 * * * *",
+				ClientIP:  strings.SplitN(node.PublicConfig.Ipv4, "/", 2)[0],
+				Bandwidth: "1M",
```
While it's correct, but I really don't like this way to construct types. Instead you should create a `NewTest` function that takes no argument (in this case) because all this stuff should be defined by the test itself, and should be modified by the test plugin (for example iperf test has specific fixed id and schedule)

Also we should have one single test for `iperf` not 2. It's a single test plugin that will do both tcp and udb (sequentially) and store the result of both tests. 

I think a task id can be only `iperf`. I don't know why it's called a Task here when it's not

muhamadazmy replied:
The tests also need to run against MANY nodes per test not a single node. I thought this is clear. so a test can take quite some time before it fiinishes (running multiple tests against multiple nodes for tcp and udp

muhamadazmy commented on pkg/network/iperf/iperf.go:
```diff
@@ -14,9 +14,18 @@ const (
+
+	return false
+}
+
```
according to my previous comment this is totally not needed

muhamadazmy commented on pkg/perf/monitor.go:
```diff
@@ -29,7 +29,7 @@ func NewPerformanceMonitor(redisAddr string) (*PerformanceMonitor, error) {
 	return &PerformanceMonitor{
 		scheduler: scheduler,
-		pool:      *redisPool,
+		pool:      redisPool,
```
revert that because i already asked Omar to fix it on his branch

muhamadazmy commented on pkg/perf/tcp_task.go:
```diff
@@ -0,0 +1,50 @@
+	TaskID    string
+	Schedule  string
+	Bandwidth string
+	ClientIP  string
```
These all need to be private fields (based on previous commet)

Also as explained, we need to run multiple tests against multiple nodes per run. hence the random list of nodes plust few fixed ones so u have like total of +10 nodes. then tests can run against these in sequence for bot tcp and udp, build and then return the report

muhamadazmy commented on pkg/perf/tcp_task.go:
```diff
@@ -0,0 +1,50 @@
+		log.Error().Err(err).Msgf("failed to run iperf tcp task: %s", stderr.String())
+	}
+
+	return out.String(), nil
```
This is not a good structure. The data from the output need to be parse and put in a structure that describe both upload and download speeds (decent structure) not just raw output from the command.

The structure should also then have information about what node, what Ip the test was run against, did it work or not, what are the speeds, and any other useful information from the test.

muhamadazmy commented on pkg/perf/udp_task.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+// UDPTask is the task for iperf udp tests
+type UDPTask struct {
```
This should become one with the other Test because u run both tcp and udp against each selected node and aggreate all results in one decent structure

rawdaGastan committed:
add iperf and graphql client

rawdaGastan committed:
merge

muhamadazmy commented on pkg/perf/iperf_task.go:
```diff
@@ -0,0 +1,123 @@
+
+// NewIperfTest creates a new iperf test
+func NewIperfTest() IperfTest {
+	return IperfTest{taskID: "iperf", schedule: "* */5 * * * *"}
```
We need to run every 6 hours according to the requirements

muhamadazmy commented on pkg/perf/iperf_task.go:
```diff
@@ -0,0 +1,123 @@
+		CpuReport:     iperfClient.Report().End.CpuReport,
+		NodeIpv4:      clientIP,
+		TestType:      string(iperfClient.Proto()),
+		Error:         err,
```
That's the wrong error, use the one from the Report

muhamadazmy commented on pkg/perf/iperf_task.go:
```diff
@@ -0,0 +1,123 @@
+	iperfClient.SetBandwidth("1M")
+	iperfClient.SetPort(iperf.IperfPort)
+	iperfClient.SetJSON(true)
+	iperfClient.SetIncludeServer(true)
```
Not sure if this is needed

muhamadazmy commented on pkg/perf/iperf_task.go:
```diff
@@ -0,0 +1,123 @@
+	var results []IperfResult
+	for _, node := range nodes {
+		clientIP := strings.SplitN(node.PublicConfig.Ipv4, "/", 2)[0]
+		clientIPv6 := strings.SplitN(node.PublicConfig.Ipv6, "/", 2)[0]
```
Possible panic because of index out of range, always validate the output that it's of the right length or use the https://pkg.go.dev/net#ParseCIDR method

muhamadazmy commented on pkg/perf/graphql_nodes.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// ListPublicNodes returns a list of public nodes
+func (g *GraphQl) ListPublicNodes(ctx context.Context, n int, farmID, excludeFarmID uint32, ipv4, ipv6 bool) ([]Node, error) {
```
what is `n`? can u give it a proper name expand the docs

rawdaGastan committed:
fix parse ips and return iperf error itself

muhamadazmy committed:
fix cron

an * in the second field means "everysecond" so it means
to run the test every second of each (6th) our of the day
